File: Sprint-1/JavaScript/findCommonItems/findCommonItems.js
/**
 * Finds common items between two arrays.
 *
 * Time Complexity: O(N*M), for each item of loop in the firstArray, it has to loop through the secondArray in the worst case
 * Space Complexity: O(N), the worst case is the length of the firstArray
 * Optimal Time Complexity: O(N + M), N is for the Set of firstArray, and M is the loop for secondArray. O(N+M) is better than O(N*M). 
 *
 * @param {Array} firstArray - First array to compare
 * @param {Array} secondArray - Second array to compare
 * @returns {Array} Array containing unique common items
 */
// export const findCommonItems = (firstArray, secondArray) => [
//   ...new Set(firstArray.filter((item) => secondArray.includes(item))),
// ];

export const findCommonItems = (firstArray, secondArray) => {
  const firstSet = new Set(firstArray);

  const commonInSecond = secondArray.filter((item) => firstSet.has(item));

  return [... new Set(commonInSecond)];
}

File: Sprint-1/Python/find_common_items/find_common_items.py
from typing import List, Sequence, TypeVar

ItemType = TypeVar("ItemType")


def find_common_items(
    first_sequence: Sequence[ItemType], second_sequence: Sequence[ItemType]
) -> List[ItemType]:
    """
    Find common items between two arrays.

    Time Complexity:
    Space Complexity:
    Optimal time complexity:
    """
    # common_items: List[ItemType] = []
    # for i in first_sequence:
    #     for j in second_sequence:
    #         if i == j and i not in common_items:
    #             common_items.append(i)
    # return common_items
def find_common_items(first_sequence, second_sequence):
    second_set = set(second_sequence)
    return list(set(first_sequence) & second_set)
